go-controller/pkg/ovn/logical_switch_manager.go

```diff
@@ -261,3 +283,131 @@ func (manager *logicalSwitchManager) ReleaseIPs(nodeName string, ipnets []*net.I
+		return allocator.NewContiguousAllocationMap(max, rangeSpec), nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return subnetRange, nil
+}
+
+// Initializes a new join switch logical switch manager
+func initJoinLogicalSwitchIPManager() (*joinSwitchIPManager, error) {
+	j := joinSwitchIPManager{
+		lsm: &logicalSwitchManager{
+			cache:    make(map[string]logicalSwitchInfo),
+			RWMutex:  sync.RWMutex{},
+			ipamFunc: NewJoinIPAMAllocator,
+		},
+		lrpIPCache:     make(map[string][]*net.IPNet),
+		lrpIPCacheLock: sync.Mutex{},
+	}
+	var joinSubnets []*net.IPNet
+	for _, joinSubnetString := range []string{v4JoinSubnetCidr, v6JoinSubnetCidr} {
+		_, joinSubnet, err := net.ParseCIDR(joinSubnetString)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing join subnet string %s: %v", joinSubnetString, err)
+		}
+		joinSubnets = append(joinSubnets, joinSubnet)
+	}
+	err := j.lsm.AddNode(ovnJoinSwitch, joinSubnets)
+	if err != nil {
+		return nil, err
+	}
+	return &j, nil
+}
+
+func (jsIPManager *joinSwitchIPManager) getJoinLRPCacheIPs(nodeName string) ([]*net.IPNet, bool) {
+	jsIPManager.lrpIPCacheLock.Lock()
+	defer jsIPManager.lrpIPCacheLock.Unlock()
+	gwLRPIPs, ok := jsIPManager.lrpIPCache[nodeName]
+	return gwLRPIPs, ok
+}
+
+func (jsIPManager *joinSwitchIPManager) setJoinLRPCacheIPs(nodeName string, gwLRPIPs []*net.IPNet) error {
+	jsIPManager.lrpIPCacheLock.Lock()
+	defer jsIPManager.lrpIPCacheLock.Unlock()
+	if _, ok := jsIPManager.lrpIPCache[nodeName]; ok {
+		return fmt.Errorf("join switch IPs already allocated for node %s", nodeName)
+	}
+	jsIPManager.lrpIPCache[nodeName] = gwLRPIPs
+	return nil
+}
+
+func (jsIPManager *joinSwitchIPManager) delJoinLRPCacheIPs(nodeName string) {
+	jsIPManager.lrpIPCacheLock.Lock()
+	defer jsIPManager.lrpIPCacheLock.Unlock()
+	delete(jsIPManager.lrpIPCache, nodeName)
+}
+
+// reserveJoinLRPIPs tries to add the LRP IPs to the joinSwitchIPManager, then they will be stored in the cache;
+func (jsIPManager *joinSwitchIPManager) reserveJoinLRPIPs(nodeName string, gwLRPIPs []*net.IPNet) (err error) {
+	// reserve the given IP in the allocator
+	if err = jsIPManager.lsm.AllocateIPs(ovnJoinSwitch, gwLRPIPs); err == nil {
+		defer func() {
+			if err != nil {
+				if relErr := jsIPManager.lsm.ReleaseIPs(ovnJoinSwitch, gwLRPIPs); relErr != nil {
+					klog.Errorf("Failed to release logical router port IPs %v just reserved for node %s: %q",
+						util.JoinIPNetIPs(gwLRPIPs, " "), nodeName, relErr)
+				}
+			}
+		}()
+		if err = jsIPManager.setJoinLRPCacheIPs(nodeName, gwLRPIPs); err != nil {
+			klog.Errorf("Failed to add reserved IPs to the join switch IP cache: %s", err.Error())
+		}
+	}
+	return err
+}
+
+// ensureJoinLRPIPs tries to allocate the LRP IPs if it is not yet allocated, then they will be stored in the cache
+func (jsIPManager *joinSwitchIPManager) ensureJoinLRPIPs(nodeName string) (gwLRPIPs []*net.IPNet, err error) {
```

Comment:
I am not sure I understand your concern. Do you see a possible race? Can you elaborate?

Reply:
I don't see any particular race, but the code doesn't make it obvious that there's _not_ a race, whereas if there was only a single lock, that would be clearer

Reply:
@danwinship Thanks. I looked it over again, and I still don't see any possible race here. Since joinSwitchIPManager.lrpIPCacheLock, logicalSwitchManager and ipam all has a lock (note logicalSwitchManager's lock is not useful here but we'd just like to reuse the existing code), I am reluctant to introduce another level of lock here unless it is necesarry.

Reply:
It looks like this is a side effect of inheriting the logical switch manager with its own locks, which then calls an ipam object which also has a lock. It looks like the code should be solid as is. I agree the more locks, the more potential pitfalls, but I don't see a good way around it with the code as is. We can perhaps refactor it in the future if it becomes a bigger problem.